File: web/frontend/lib/core/terminal_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import '../core/auth.dart';

enum TerminalConnectionState { disconnected, connecting, connected, error }

/// Result of syncing env vars to the gateway config.
class EnvSyncResult {
  final List<String> synced;
  final List<String> skipped;
  final List<String> errors;
  final String message;

  const EnvSyncResult({
    required this.synced,
    required this.skipped,
    required this.errors,
    required this.message,
  });
}

class TerminalOutput {
  final String type; // 'stdout', 'stderr', 'system', 'error', 'exit'
  final String? data;
  final String? message;
  final int? exitCode;
  final DateTime timestamp;

  TerminalOutput({
    required this.type,
    this.data,
    this.message,
    this.exitCode,
    DateTime? timestamp,
  }) : timestamp = timestamp ?? DateTime.now();
}

class TerminalProxyClient extends ChangeNotifier {
  final String url;
  final GatewayAuth auth;
  String _role;

  static const int _maxOutputs = 10000;

  WebSocketChannel? _channel;
  TerminalConnectionState _state = TerminalConnectionState.disconnected;
  final List<TerminalOutput> _outputs = [];
  bool _isAuthenticated = false;
  bool _isExecuting = false;
  bool _disposed = false;
  int _reconnectAttempts = 0;
  Timer? _reconnectTimer;

  /// Completer that resolves when auth succeeds (or fails/times out).
  Completer<void>? _authCompleter;

  /// Completers for pending env operations (resolved in _onMessage).
  Completer<Map<String, String>>? _envListCompleter;
  Completer<void>? _envSetCompleter;
  Completer<void>? _envDeleteCompleter;
  Completer<EnvSyncResult>? _envSyncCompleter;

  /// Command queue: each entry is a Future that completes when the previous
  /// command finishes. This ensures only one command runs at a time even when
  /// multiple callers invoke executeCommandForOutput concurrently.
  Future<void> _execQueue = Future.value();

  TerminalConnectionState get state => _state;
  List<TerminalOutput> get outputs => List.unmodifiable(_outputs);
  bool get isAuthenticated => _isAuthenticated;
  bool get isExecuting => _isExecuting;
  bool get isConnected => _state == TerminalConnectionState.connected;

  TerminalProxyClient({
    required this.url,
    required this.auth,
    String role = 'admin',
  }) : _role = role;

  /// Update the role used for terminal proxy authentication.
  void updateRole(String role) {
    _role = role;
  }

  /// Connect to the terminal proxy and wait for authentication to complete.
  /// Returns when authenticated, or throws on auth failure / timeout.
  Future<void> connect() async {
    // Already connected and authed -- nothing to do.
    if (_state == TerminalConnectionState.connected && _isAuthenticated) return;

    // If already connecting, wait for the in-flight auth to finish.
    if (_state == TerminalConnectionState.connecting && _authCompleter != null) {
      return _authCompleter!.future;
    }

    _state = TerminalConnectionState.connecting;
    _authCompleter = Completer<void>();
    notifyListeners();

    try {
      _channel = WebSocketChannel.connect(Uri.parse(url));
      await _channel!.ready; // Wait for WebSocket handshake to complete

      _channel!.stream.listen(
        _onMessage,
        onError: _onError,
        onDone: _onDone,
      );

      _authenticate();

      // Wait for the auth response (resolved in _onMessage 'auth' case).
      // Timeout after 10 seconds so callers don't hang forever.
      await _authCompleter!.future.timeout(
        const Duration(seconds: 10),
        onTimeout: () {
          if (!_authCompleter!.isCompleted) {
            _authCompleter!.completeError(
              TimeoutException('Terminal proxy auth timed out'),
            );
          }
        },
      );
    } catch (e) {
      _state = TerminalConnectionState.error;
      notifyListeners();
      rethrow;
    }
  }

  void _authenticate() {
    _channel?.sink.add(jsonEncode({
      'type': 'auth',
      'token': auth.token,
      'role': _role,
    }));
  }

  void _addOutput(TerminalOutput output) {
    _outputs.add(output);
    // Evict oldest entries if over capacity
    if (_outputs.length > _maxOutputs) {
      _outputs.removeRange(0, _outputs.length - _maxOutputs);
    }
  }

  void _onMessage(dynamic raw) {
    try {
      final data = jsonDecode(raw as String);
      final type = data['type'] as String;

      switch (type) {
        case 'auth':
          _isAuthenticated = data['status'] == 'ok';
          if (_isAuthenticated) {
            _state = TerminalConnectionState.connected;
            _reconnectAttempts = 0;
            _reconnectTimer?.cancel();
            if (_authCompleter != null && !_authCompleter!.isCompleted) {
              _authCompleter!.complete();
            }
          } else {
            _state = TerminalConnectionState.error;
            if (_authCompleter != null && !_authCompleter!.isCompleted) {
              _authCompleter!.completeError(
                StateError('Terminal proxy auth failed'),
              );
            }
          }
          notifyListeners();
          break;

        case 'stdout':
          _addOutput(TerminalOutput(
            type: 'stdout',
            data: data['data'] as String,
          ));
          notifyListeners();
          break;

        case 'stderr':
          _addOutput(TerminalOutput(
            type: 'stderr',
            data: data['data'] as String,
          ));
          notifyListeners();
          break;

        case 'system':
          _addOutput(TerminalOutput(
            type: 'system',
            message: data['message'] as String,
          ));
          notifyListeners();
          break;

        case 'error':
          _isExecuting = false;
          final errMsg = data['message'] as String;
          _addOutput(TerminalOutput(
            type: 'error',
            message: errMsg,
          ));
          // Fail any pending env completers -- the server sends generic
          // {type:'error'} for auth/permission failures on env operations
          // instead of a typed env_* response.
          _failPendingEnvCompleters(errMsg);
          notifyListeners();
          break;

        case 'exit':
          _isExecuting = false;
          _addOutput(TerminalOutput(
            type: 'exit',
            exitCode: data['code'] as int?,
            message: data['message'] as String?,
          ));
          notifyListeners();
          break;

        case 'pong':
          // Keep alive response
          break;

        // ── Dynamic env var responses ──────────────────────────────────
        case 'env_list':
          if (_envListCompleter != null && !_envListCompleter!.isCompleted) {
            final rawVars = data['vars'] as Map<String, dynamic>? ?? {};
            final vars = rawVars.map((k, v) => MapEntry(k, v.toString()));
            _envListCompleter!.complete(vars);
          }
          break;

        case 'env_set':
          if (data['status'] == 'ok') {
            if (_envSetCompleter != null && !_envSetCompleter!.isCompleted) {
              _envSetCompleter!.complete();
            }
          } else {
            final msg = data['message'] as String? ?? 'Failed to set env var';
            if (_envSetCompleter != null && !_envSetCompleter!.isCompleted) {
              _envSetCompleter!.completeError(StateError(msg));
            }
          }
          break;

        case 'env_delete':
          if (data['status'] == 'ok') {
            if (_envDeleteCompleter != null && !_envDeleteCompleter!.isCompleted) {
              _envDeleteCompleter!.complete();
            }
          } else {
            final msg = data['message'] as String? ?? 'Failed to delete env var';
            if (_envDeleteCompleter != null && !_envDeleteCompleter!.isCompleted) {
              _envDeleteCompleter!.completeError(StateError(msg));
            }
          }
          break;

        case 'env_sync_gateway':
          if (_envSyncCompleter != null && !_envSyncCompleter!.isCompleted) {
            final syncedRaw = data['synced'] as List<dynamic>? ?? [];
            final skippedRaw = data['skipped'] as List<dynamic>? ?? [];
            final errorsRaw = data['errors'] as List<dynamic>? ?? [];
            final result = EnvSyncResult(
              synced: syncedRaw.map((e) => e.toString()).toList(),
              skipped: skippedRaw.map((e) => e.toString()).toList(),
              errors: errorsRaw.map((e) => e.toString()).toList(),
              message: data['message'] as String? ?? '',
            );
            if (data['status'] == 'ok') {
              _envSyncCompleter!.complete(result);
            } else {
              // Still provide the result via completeError wrapping
              _envSyncCompleter!.completeError(result);
            }
          }
          break;
      }
    } catch (e) {
      if (kDebugMode) debugPrint('[Terminal] Error parsing message: $e');
    }
  }

  void _onError(dynamic error) {
    _state = TerminalConnectionState.error;
    _outputs.add(TerminalOutput(
      type: 'error',
      message: 'Connection error: $error',
    ));
    if (_authCompleter != null && !_authCompleter!.isCompleted) {
      _authCompleter!.completeError(error);
    }
    _failPendingEnvCompleters('Connection error');
    notifyListeners();
  }

  void _onDone() {
    _state = TerminalConnectionState.disconnected;
    _isAuthenticated = false;
    if (_authCompleter != null && !_authCompleter!.isCompleted) {
      _authCompleter!.completeError(
        StateError('Terminal proxy connection closed before auth'),
      );
    }
    _failPendingEnvCompleters('Connection closed');
    notifyListeners();
    _scheduleReconnect();
  }

  /// Auto-reconnect with exponential backoff (mirrors GatewayClient pattern)
  void _scheduleReconnect() {
    if (_disposed) return;
    _reconnectAttempts++;
    // Exponential backoff: 1s, 2s, 4s, 8s, 16s, 30s max
    final delay = Duration(
        seconds: (_reconnectAttempts <= 5)
            ? (1 << (_reconnectAttempts - 1))
            : 30);
    if (kDebugMode) debugPrint('[Terminal] reconnecting in ${delay.inSeconds}s (attempt $_reconnectAttempts)');
    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(delay, () {
      if (!_disposed &&
          _state != TerminalConnectionState.connected &&
          _state != TerminalConnectionState.connecting) {
        connect().catchError((_) {});
      }
    });
  }

  void executeCommand(String command) {
    if (!_isAuthenticated || _channel == null) {
      _outputs.add(TerminalOutput(
        type: 'error',
        message: 'Not connected to terminal proxy',
      ));
      notifyListeners();
      return;
    }

    _isExecuting = true;
    _channel!.sink.add(jsonEncode({
      'type': 'exec',
      'command': command,
    }));
    notifyListeners();
  }

  /// Execute a command and return its collected stdout as a string.
  ///
  /// Commands are queued: if multiple callers invoke this concurrently, each
  /// command waits for the previous one to finish before starting. This
  /// prevents output interleaving on the single WebSocket connection.
  Future<String> executeCommandForOutput(
    String command, {
    Duration timeout = const Duration(seconds: 30),
  }) {
    // Chain onto the queue so only one command runs at a time.
    final previous = _execQueue;
    final completer = Completer<String>();
    _execQueue = completer.future.catchError((_) {});
    // ignore errors in the queue chain so a failed command doesn't block the next

    previous.then((_) {
      _executeCommandSingle(command, timeout: timeout).then(
        (result) { if (!completer.isCompleted) completer.complete(result); },
        onError: (e) { if (!completer.isCompleted) completer.completeError(e); },
      );
    });

    return completer.future;
  }

  /// Internal: run a single command after the queue has released.
  Future<String> _executeCommandSingle(
    String command, {
    Duration timeout = const Duration(seconds: 30),
  }) async {
    if (!_isAuthenticated || _channel == null) {
      throw StateError('Not connected to terminal proxy');
    }

    // If somehow still executing (e.g. stale state), wait for it to clear.
    if (_isExecuting) {
      final waitCompleter = Completer<void>();
      Timer? waitTimer;

      void waitListener() {
        if (!_isExecuting && !waitCompleter.isCompleted) {
          removeListener(waitListener);
          waitTimer?.cancel();
          waitCompleter.complete();
        }
      }

      addListener(waitListener);
      waitTimer = Timer(timeout, () {
        removeListener(waitListener);
        if (!waitCompleter.isCompleted) {
          waitCompleter.completeError(
            TimeoutException('Timed out waiting for previous command to finish', timeout),
          );
        }
      });

      await waitCompleter.future;
    }

    final startIndex = _outputs.length;
    final resultCompleter = Completer<String>();
    Timer? timer;

    void finish() {
      if (resultCompleter.isCompleted) return;
      final slice = _outputs.sublist(startIndex);
      final buffer = StringBuffer();
      for (final out in slice) {
        final text = out.data ?? out.message;
        if (text == null || text.isEmpty) continue;
        buffer.writeln(text);
      }
      resultCompleter.complete(buffer.toString());
    }

    void listener() {
      if (!_isExecuting) {
        removeListener(listener);
        timer?.cancel();
        finish();
      }
    }

    addListener(listener);
    timer = Timer(timeout, () {
      removeListener(listener);
      if (!resultCompleter.isCompleted) {
        resultCompleter.completeError(
          TimeoutException('Command timed out: $command', timeout),
        );
      }
    });

    executeCommand(command);
    return resultCompleter.future;
  }

  void cancelCommand() {
    if (_channel != null) {
      _channel!.sink.add(jsonEncode({
        'type': 'cancel',
      }));
    }
  }

  // ── Dynamic environment variable management (superadmin only) ─────────

  /// Fail all pending env completers with the given error message.
  /// Called when the connection drops or the server sends a generic error.
  void _failPendingEnvCompleters(String reason) {
    final err = StateError(reason);
    if (_envListCompleter != null && !_envListCompleter!.isCompleted) {
      _envListCompleter!.completeError(err);
    }
    if (_envSetCompleter != null && !_envSetCompleter!.isCompleted) {
      _envSetCompleter!.completeError(err);
    }
    if (_envDeleteCompleter != null && !_envDeleteCompleter!.isCompleted) {
      _envDeleteCompleter!.completeError(err);
    }
    if (_envSyncCompleter != null && !_envSyncCompleter!.isCompleted) {
      _envSyncCompleter!.completeError(err);
    }
  }

  /// List all dynamic env var overrides.
  Future<Map<String, String>> listEnvVars({
    Duration timeout = const Duration(seconds: 5),
  }) async {
    if (!_isAuthenticated || _channel == null) {
      throw StateError('Not connected to terminal proxy');
    }
    // Cancel any previous in-flight request to prevent orphaned completers
    if (_envListCompleter != null && !_envListCompleter!.isCompleted) {
      _envListCompleter!.completeError(StateError('Superseded by new request'));
    }
    _envListCompleter = Completer<Map<String, String>>();
    _channel!.sink.add(jsonEncode({'type': 'env_list'}));
    return _envListCompleter!.future.timeout(timeout);
  }

  /// Set or update a dynamic env var override.
  Future<void> setEnvVar(
    String key,
    String value, {
    Duration timeout = const Duration(seconds: 5),
  }) async {
    if (!_isAuthenticated || _channel == null) {
      throw StateError('Not connected to terminal proxy');
    }
    if (_envSetCompleter != null && !_envSetCompleter!.isCompleted) {
      _envSetCompleter!.completeError(StateError('Superseded by new request'));
    }
    _envSetCompleter = Completer<void>();
    _channel!.sink.add(jsonEncode({
      'type': 'env_set',
      'key': key,
      'value': value,
    }));
    return _envSetCompleter!.future.timeout(timeout);
  }

  /// Delete a dynamic env var override.
  Future<void> deleteEnvVar(
    String key, {
    Duration timeout = const Duration(seconds: 5),
  }) async {
    if (!_isAuthenticated || _channel == null) {
      throw StateError('Not connected to terminal proxy');
    }
    if (_envDeleteCompleter != null && !_envDeleteCompleter!.isCompleted) {
      _envDeleteCompleter!.completeError(StateError('Superseded by new request'));
    }
    _envDeleteCompleter = Completer<void>();
    _channel!.sink.add(jsonEncode({
      'type': 'env_delete',
      'key': key,
    }));
    return _envDeleteCompleter!.future.timeout(timeout);
  }

  /// Sync all env overrides with known config mappings into the gateway
  /// config file, then trigger a gateway restart.
  Future<EnvSyncResult> syncEnvToGateway({
    Duration timeout = const Duration(seconds: 30),
  }) async {
    if (!_isAuthenticated || _channel == null) {
      throw StateError('Not connected to terminal proxy');
    }
    if (_envSyncCompleter != null && !_envSyncCompleter!.isCompleted) {
      _envSyncCompleter!.completeError(StateError('Superseded by new request'));
    }
    _envSyncCompleter = Completer<EnvSyncResult>();
    _channel!.sink.add(jsonEncode({'type': 'env_sync_gateway'}));
    return _envSyncCompleter!.future.timeout(timeout);
  }

  void clearOutput() {
    _outputs.clear();
    notifyListeners();
  }

  void disconnect() {
    _reconnectTimer?.cancel();
    _reconnectAttempts = 0;
    _channel?.sink.close();
    _state = TerminalConnectionState.disconnected;
    _isAuthenticated = false;
    _isExecuting = false;
    if (_authCompleter != null && !_authCompleter!.isCompleted) {
      _authCompleter!.completeError(
        StateError('Terminal proxy disconnected'),
      );
    }
    _authCompleter = null;
    _failPendingEnvCompleters('Disconnected');
    notifyListeners();
  }

  @override
  void dispose() {
    _disposed = true;
    disconnect();
    super.dispose();
  }

  void sendPing() {
    _channel?.sink.add(jsonEncode({'type': 'ping'}));
  }
}

File: app/web/src/lib/clients/terminal-client.ts
/**
 * TerminalProxyClient — 1:1 port of core/terminal_client.dart
 *
 * WebSocket client for the terminal proxy service.
 * Auth, command execution, env var management, interactive PTY shell.
 */

/* ------------------------------------------------------------------ */
/*  Types                                                              */
/* ------------------------------------------------------------------ */

export type TerminalConnectionState = 'disconnected' | 'connecting' | 'connected' | 'error';

export interface TerminalOutput {
  type: 'stdout' | 'stderr' | 'system' | 'error' | 'exit';
  data?: string;
  message?: string;
  exitCode?: number;
  timestamp: number;
}

export interface EnvSyncResult {
  synced: number;
  skipped: number;
  errors: number;
  message: string;
}

export type TerminalEventListener = (output: TerminalOutput) => void;
export type ShellOutputListener = (data: string) => void;

/* ------------------------------------------------------------------ */
/*  TerminalProxyClient                                                */
/* ------------------------------------------------------------------ */

export class TerminalProxyClient {
  private _wsUrl: string;
  private _token: string;
  private _role: string;
  private _openClawId: string | null = null;
  private _ws: WebSocket | null = null;
  private _connectionState: TerminalConnectionState = 'disconnected';
  private _isAuthenticated = false;
  private _authCompleter: { resolve: () => void; reject: (e: Error) => void } | null = null;
  private _connectPromise: Promise<void> | null = null;
  private _reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private _reconnectAttempt = 0;

  // Outputs
  private _outputs: TerminalOutput[] = [];
  private _maxOutputs = 10000;
  private _outputListeners = new Set<TerminalEventListener>();
  private _stateListeners = new Set<(state: TerminalConnectionState) => void>();

  // Shell
  private _shellActive = false;
  private _shellOutputListeners = new Set<ShellOutputListener>();

  // Exec queue
  private _execQueue: Array<{
    command: string;
    resolve: (output: string) => void;
    reject: (err: Error) => void;
    timeout: number;
  }> = [];
  private _execRunning = false;

  // Env completers
  private _envListCompleter: { resolve: (v: Record<string, string>) => void; reject: (e: Error) => void } | null = null;
  private _envSetCompleter: { resolve: () => void; reject: (e: Error) => void } | null = null;
  private _envDeleteCompleter: { resolve: () => void; reject: (e: Error) => void } | null = null;
  private _envSyncCompleter: { resolve: (v: EnvSyncResult) => void; reject: (e: Error) => void } | null = null;

  constructor(wsUrl: string, token: string, role: string) {
    this._wsUrl = wsUrl;
    this._token = token;
    this._role = role;
  }

  get connectionState(): TerminalConnectionState { return this._connectionState; }
  get isAuthenticated(): boolean { return this._isAuthenticated; }
  get outputs(): TerminalOutput[] { return this._outputs; }
  get shellActive(): boolean { return this._shellActive; }

  set openClawId(id: string | null) { this._openClawId = id; }
  set token(t: string) { this._token = t; }

  /* ---------------------------------------------------------------- */
  /*  Subscriptions                                                    */
  /* ---------------------------------------------------------------- */

  onOutput(fn: TerminalEventListener): () => void {
    this._outputListeners.add(fn);
    return () => this._outputListeners.delete(fn);
  }

  onShellOutput(fn: ShellOutputListener): () => void {
    this._shellOutputListeners.add(fn);
    return () => this._shellOutputListeners.delete(fn);
  }

  onStateChange(fn: (state: TerminalConnectionState) => void): () => void {
    this._stateListeners.add(fn);
    return () => this._stateListeners.delete(fn);
  }

  private _setConnectionState(state: TerminalConnectionState): void {
    this._connectionState = state;
    this._stateListeners.forEach((fn) => fn(state));
  }

  private _addOutput(output: TerminalOutput): void {
    this._outputs.push(output);
    if (this._outputs.length > this._maxOutputs) {
      this._outputs = this._outputs.slice(-this._maxOutputs);
    }
    this._outputListeners.forEach((fn) => fn(output));
  }

  /* ---------------------------------------------------------------- */
  /*  Connect / disconnect                                             */
  /* ---------------------------------------------------------------- */

  connect(): Promise<void> {
    if (this._connectPromise) return this._connectPromise;
    if (this._connectionState === 'connected') return Promise.resolve();

    this._setConnectionState('connecting');

    this._connectPromise = new Promise<void>((resolve, reject) => {
      try {
        this._ws = new WebSocket(this._wsUrl);
      } catch {
        this._setConnectionState('error');
        this._connectPromise = null;
        reject(new Error('Failed to create WebSocket'));
        return;
      }

      this._ws.onopen = () => {
        // Send auth message
        this._send({
          type: 'auth',
          jwt: this._token,
          role: this._role,
          openclawId: this._openClawId,
        });

        // Wait for auth response (10s timeout)
        const timer = setTimeout(() => {
          if (!this._isAuthenticated) {
            this._authCompleter?.reject(new Error('Auth timeout'));
            this._authCompleter = null;
            this._connectPromise = null;
            this._ws?.close();
          }
        }, 10000);

        this._authCompleter = {
          resolve: () => {
            clearTimeout(timer);
            this._isAuthenticated = true;
            this._setConnectionState('connected');
            this._reconnectAttempt = 0;
            this._connectPromise = null;
            resolve();
          },
          reject: (e: Error) => {
            clearTimeout(timer);
            this._connectPromise = null;
            reject(e);
          },
        };
      };

      this._ws.onmessage = (e) => this._onMessage(String(e.data));

      this._ws.onclose = () => {
        this._isAuthenticated = false;
        this._shellActive = false;
        this._setConnectionState('disconnected');
        this._connectPromise = null;
        this._scheduleReconnect();
      };

      this._ws.onerror = () => {
        this._setConnectionState('error');
        this._connectPromise = null;
        reject(new Error('WebSocket error'));
      };
    });

    return this._connectPromise;
  }

  disconnect(): void {
    this._clearReconnect();
    this._connectPromise = null;
    if (this._ws) {
      try {
        this._ws.close();
      } catch {
        // Ignore
      }
      this._ws = null;
    }
    this._isAuthenticated = false;
    this._shellActive = false;
    this._setConnectionState('disconnected');
  }

  private _send(data: any): void {
    if (this._ws?.readyState === WebSocket.OPEN) {
      this._ws.send(JSON.stringify(data));
    }
  }

  /* ---------------------------------------------------------------- */
  /*  Message handling                                                 */
  /* ---------------------------------------------------------------- */

  private _onMessage(raw: string): void {
    let msg: any;
    try {
      msg = JSON.parse(raw);
    } catch {
      return;
    }

    switch (msg.type) {
      case 'auth':
        this._authCompleter?.resolve();
        this._authCompleter = null;
        break;

      case 'stdout':
      case 'stderr':
      case 'system':
        this._addOutput({ type: msg.type, data: msg.data ?? msg.message, timestamp: Date.now() });
        break;

      case 'error':
        this._addOutput({ type: 'error', message: msg.message ?? msg.error, timestamp: Date.now() });
        this._envSetCompleter?.reject(new Error(msg.message));
        this._envSetCompleter = null;
        this._envDeleteCompleter?.reject(new Error(msg.message));
        this._envDeleteCompleter = null;
        break;

      case 'exit':
        this._addOutput({ type: 'exit', exitCode: msg.exitCode ?? msg.code, timestamp: Date.now() });
        this._execRunning = false;
        this._processExecQueue();
        break;

      case 'pong':
        break;

      case 'env_list':
        this._envListCompleter?.resolve(msg.vars ?? {});
        this._envListCompleter = null;
        break;

      case 'env_set':
        if (msg.status === 'ok') this._envSetCompleter?.resolve();
        else this._envSetCompleter?.reject(new Error(msg.error ?? 'Failed'));
        this._envSetCompleter = null;
        break;

      case 'env_delete':
        if (msg.status === 'ok') this._envDeleteCompleter?.resolve();
        else this._envDeleteCompleter?.reject(new Error(msg.error ?? 'Failed'));
        this._envDeleteCompleter = null;
        break;

      case 'env_sync_gateway':
        this._envSyncCompleter?.resolve({
          synced: msg.synced ?? 0,
          skipped: msg.skipped ?? 0,
          errors: msg.errors ?? 0,
          message: msg.message ?? '',
        });
        this._envSyncCompleter = null;
        break;

      case 'shell_started':
        this._shellActive = true;
        break;

      case 'shell_output':
        this._shellOutputListeners.forEach((fn) => fn(msg.data ?? ''));
        break;

      case 'shell_closed':
        this._shellActive = false;
        break;
    }
  }

  /* ---------------------------------------------------------------- */
  /*  Command execution                                                */
  /* ---------------------------------------------------------------- */

  executeCommand(command: string): void {
    this._send({ type: 'exec', command });
  }

  executeCommandForOutput(command: string, timeout = 30000): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      this._execQueue.push({ command, resolve, reject, timeout });
      this._processExecQueue();
    });
  }

  private _processExecQueue(): void {
    if (this._execRunning || this._execQueue.length === 0) return;
    const next = this._execQueue.shift()!;
    this._execRunning = true;

    const startIdx = this._outputs.length;
    this.executeCommand(next.command);

    const timer = setTimeout(() => {
      this._execRunning = false;
      next.resolve(this._collectOutput(startIdx));
      this._processExecQueue();
    }, next.timeout);

    // Watch for exit event
    const unsub = this.onOutput((output) => {
      if (output.type === 'exit') {
        clearTimeout(timer);
        unsub();
        // Give a small delay for final output
        setTimeout(() => {
          next.resolve(this._collectOutput(startIdx));
        }, 50);
      }
    });
  }

  private _collectOutput(fromIdx: number): string {
    return this._outputs
      .slice(fromIdx)
      .filter((o) => o.type === 'stdout' || o.type === 'stderr')
      .map((o) => o.data ?? o.message ?? '')
      .join('\n');
  }

  cancelCommand(): void {
    this._send({ type: 'cancel' });
  }

  /* ---------------------------------------------------------------- */
  /*  Env var management                                               */
  /* ---------------------------------------------------------------- */

  listEnvVars(): Promise<Record<string, string>> {
    return new Promise((resolve, reject) => {
      this._envListCompleter = { resolve, reject };
      this._send({ type: 'env_list' });
      setTimeout(() => {
        if (this._envListCompleter) {
          this._envListCompleter.reject(new Error('Timeout'));
          this._envListCompleter = null;
        }
      }, 5000);
    });
  }

  setEnvVar(key: string, value: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this._envSetCompleter = { resolve, reject };
      this._send({ type: 'env_set', key, value });
      setTimeout(() => {
        if (this._envSetCompleter) {
          this._envSetCompleter.reject(new Error('Timeout'));
          this._envSetCompleter = null;
        }
      }, 5000);
    });
  }

  deleteEnvVar(key: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this._envDeleteCompleter = { resolve, reject };
      this._send({ type: 'env_delete', key });
      setTimeout(() => {
        if (this._envDeleteCompleter) {
          this._envDeleteCompleter.reject(new Error('Timeout'));
          this._envDeleteCompleter = null;
        }
      }, 5000);
    });
  }

  syncEnvToGateway(): Promise<EnvSyncResult> {
    return new Promise((resolve, reject) => {
      this._envSyncCompleter = { resolve, reject };
      this._send({ type: 'env_sync_gateway' });
      setTimeout(() => {
        if (this._envSyncCompleter) {
          this._envSyncCompleter.reject(new Error('Timeout'));
          this._envSyncCompleter = null;
        }
      }, 30000);
    });
  }

  /* ---------------------------------------------------------------- */
  /*  Interactive PTY shell                                            */
  /* ---------------------------------------------------------------- */

  startShell(cols: number, rows: number): void {
    this._send({ type: 'shell_start', cols, rows });
  }

  shellInput(data: string): void {
    this._send({ type: 'shell_input', data });
  }

  shellResize(cols: number, rows: number): void {
    this._send({ type: 'shell_resize', cols, rows });
  }

  closeShell(): void {
    this._send({ type: 'shell_close' });
  }

  /* ---------------------------------------------------------------- */
  /*  Reconnection                                                     */
  /* ---------------------------------------------------------------- */

  private _scheduleReconnect(): void {
    this._clearReconnect();
    const delay = Math.min(1000 * Math.pow(2, this._reconnectAttempt), 30000);
    this._reconnectAttempt++;
    this._reconnectTimer = setTimeout(() => this.connect(), delay);
  }

  private _clearReconnect(): void {
    if (this._reconnectTimer) {
      clearTimeout(this._reconnectTimer);
      this._reconnectTimer = null;
    }
  }
}
